faults/common.go

```diff
@@ -0,0 +1,124 @@
+// Package faults provides common headers and client-side fault injection functionality.
+package faults
+
+import (
+	"fmt"
+	"log/slog"
+	"math/rand/v2"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type GetHeaderFn func(key string) string
+type ResumeFn func() (interface{}, error)
+type ResponseFn func(code int, message string) (interface{}, error)
+type SleepFn func(d time.Duration)
+
+// Object to ensure a random number is only generated at most 1 time.
+type randSingleton struct {
+	randInt *int
+}
+
+func (r randSingleton) getRandInt() int {
+	if r.randInt == nil {
+		*r.randInt = rand.IntN(100)
+	}
+	return *r.randInt
+}
+
+func isSelected(percentageHeader string, GetHeaderFn func(key string) string, singleRand randSingleton) (bool, string) {
+	percentageStr := GetHeaderFn(percentageHeader)
+	if percentageStr == "" {
+		return true, ""
+	}
+	percentage, err := strconv.Atoi(percentageStr)
+	if err != nil {
+		return false, fmt.Sprintf("provided percentage \"%s\" is not a valid integer", percentageStr)
+	}
+	if percentage < 0 || percentage > 100 {
+		return false, fmt.Sprintf("provided percentage \"%d\" is outside the valid range of [0-100]", percentage)
+	}
+	return singleRand.getRandInt() < percentage, ""
+}
+
+type InjectFaultParams struct {
+	CallerName string
+
+	Address, Method            string
+	AbortCodeMin, AbortCodeMax int
+
+	GetHeaderFn GetHeaderFn
+	ResumeFn    ResumeFn
+	ResponseFn  ResponseFn
+
+	// Exposed for tests.
+	RandInt *int
+	SleepFn *SleepFn
+}
+
+func InjectFault(params InjectFaultParams) (interface{}, error) {
+	serverAddress := params.GetHeaderFn(FaultServerAddressHeader)
+	shortAddress := params.Address
+	if i := strings.Index(params.Address, ".svc.cluster.local"); i != -1 {
+		shortAddress = params.Address[:i]
+	}
```

Comment:
use [`strings.CutSuffix`](https://pkg.go.dev/strings#CutSuffix) instead?

your current implementation will also cut it if there's something after `.svc.cluster.local`, I'm not sure if that's intentional or not (what could be after it and do we really want to cut it in that case?)

Reply:
The current implementation is correct: Everything including and after `.svc.cluster.local` should be removed. `CutSuffix` doesn't allow for this functionality when the values after `.svc.cluster.local` are not known -- i.e. port values. I added a comment to clarify this.